clients/src/test/java/org/apache/kafka/common/security/authenticator/SaslServerAuthenticatorTest.java

```diff
@@ -92,7 +102,82 @@ public void testUnexpectedRequestType() throws IOException {
         verify(transportLayer, times(2)).read(any(ByteBuffer.class));
     }
 
-    private SaslServerAuthenticator setupAuthenticator(Map<String, ?> configs, TransportLayer transportLayer, String mechanism) throws IOException {
+    @Test
+    public void testOldestApiVersionsRequest() throws IOException {
+        final short version = ApiKeys.API_VERSIONS.oldestVersion();
+        TransportLayer transportLayer = mock(TransportLayer.class, Answers.RETURNS_DEEP_STUBS);
+        Map<String, ?> configs = Collections.singletonMap(BrokerSecurityConfigs.SASL_ENABLED_MECHANISMS_CONFIG,
+            Collections.singletonList(SCRAM_SHA_256.mechanismName()));
+        ChannelMetadataRegistry metadataRegistry = new DefaultChannelMetadataRegistry();
+        SaslServerAuthenticator authenticator = setupAuthenticator(configs, transportLayer,
+            SCRAM_SHA_256.mechanismName(), metadataRegistry);
+
+        final RequestHeader header = new RequestHeader(ApiKeys.API_VERSIONS, version, "clientId", 0);
+        final Struct headerStruct = header.toStruct();
+
+        final ApiVersionsRequest request = new ApiVersionsRequest.Builder().build(version);
+        final Struct requestStruct = request.data.toStruct(version);
+
+        when(transportLayer.socketChannel().socket().getInetAddress()).thenReturn(InetAddress.getLoopbackAddress());
+
+        when(transportLayer.read(any(ByteBuffer.class))).then(invocation -> {
+            invocation.<ByteBuffer>getArgument(0).putInt(headerStruct.sizeOf() + requestStruct.sizeOf());
+            return 4;
+        }).then(invocation -> {
+            headerStruct.writeTo(invocation.getArgument(0));
+            requestStruct.writeTo(invocation.getArgument(0));
+            return headerStruct.sizeOf() + requestStruct.sizeOf();
+        });
+
+        authenticator.authenticate();
+
+        assertEquals(ClientInformation.UNKNOWN_NAME_OR_VERSION,
+            metadataRegistry.clientInformation().softwareName());
+        assertEquals(ClientInformation.UNKNOWN_NAME_OR_VERSION,
+            metadataRegistry.clientInformation().softwareVersion());
+
+        verify(transportLayer, times(2)).read(any(ByteBuffer.class));
+    }
+
+    @Test
+    public void testLatestApiVersionsRequest() throws IOException {
```

Comment:
It seems like this is very similar to the previous test case, but with a different version and it returns the actual info rather than "unknown".  Any chance we could share some code?

Reply:
Definitely.